### PR TITLE
Fill in team icon upload gql API

### PR DESCRIPTION
**Pull request by duanxianbo**

### CODE REVIEW

File: core/service/team.go [hunk 1]
```diff
@@ -2,16 +2,25 @@ package service
...
 type Team struct {
-	taskDao   dao.Task
-	sprintDao dao.Sprint
-	teamDao   dao.Team
+	cloudWebAPIExternalBaseURL string
+	taskDao                    dao.Task
+	sprintDao                  dao.Sprint
+	teamDao                    dao.Team
+	cloudClientRegistry        *cloudAPI.ClientRegistry
+	teamFileUploadSessionDao   dao.TeamFileUploadSession
```
Comment (naturetime10) [inline core/service/team.go]:
```suggestion
	cloudClientRegistry        *cloudAPI.ClientRegistry
	taskDao                    dao.Task
	sprintDao                  dao.Sprint
	teamDao                    dao.Team
	teamFileUploadSessionDao   dao.TeamFileUploadSession
```

File: core/service/team.go [hunk 2]
```diff
@@ -56,6 +65,87 @@ func (t Team) FindSprintsInTeam(ct context.Context, teamID uint64, filter *Sprin
...
+	err = t.teamFileUploadSessionDao.CreateTeamFileUploadSession(fileUploadSession)
+	return res.UploadSessionId, err
+}
+
+func (t Team) FinishTeamIconUploadSession(ct context.Context, teamID uint64, fileUploadSessionID uint64) (uint64, error) {
+	iconUploadSession, err := t.teamFileUploadSessionDao.FindTeamFileUploadSessionByTeamID(
+		teamID,
+		entity.IconTeamFileUploadSessionType,
+		fileUploadSessionID)
+
```
Comment (naturetime10) [inline core/service/team.go]:
```suggestion
```

File: core/service/team.go [hunk 3]
```diff
@@ -56,6 +65,87 @@ func (t Team) FindSprintsInTeam(ct context.Context, teamID uint64, filter *Sprin
...
+	if err != nil {
+		log.Println(err)
+		return 0, err
+	}
+
+	if iconUploadSession.IsCompleted {
+		err = fmt.Errorf("profile upload session is already completed: teamID=%v, fileUploadSessionID=%v",
```
Comment (naturetime10) [inline core/service/team.go]:
```suggestion
		err = fmt.Errorf("icon upload session is already completed: teamID=%v, fileUploadSessionID=%v",
```

File: core/service/team.go [hunk 4]
```diff
@@ -56,6 +65,87 @@ func (t Team) FindSprintsInTeam(ct context.Context, teamID uint64, filter *Sprin
...
+	err = t.teamFileUploadSessionDao.UpdateTeamFileUploadSession(iconUploadSession)
+	if err != nil {
+		log.Println(err)
+		return 0, err
+	}
+
+	findUploadSessionReq := proto.FindUploadSessionRequest{
+		UploadSessionId: fileUploadSessionID,
+	}
+
```
Comment (naturetime10) [inline core/service/team.go]:
```suggestion
```

File: core/service/team.go [hunk 5]
```diff
@@ -56,6 +65,87 @@ func (t Team) FindSprintsInTeam(ct context.Context, teamID uint64, filter *Sprin
...
+	if err != nil {
+		log.Println(err)
+		return 0, err
+	}
+
+	iconUrl := io.GetFileURL(t.cloudWebAPIExternalBaseURL, uploadSession.FileId)
+	team.IconURL = &iconUrl
+	team.UpdatedAt = &now
+	err = t.teamDao.UpdateTeam(team)
+	return fileUploadSessionID, err
```
Comment (naturetime10) [inline core/service/team.go]:
```suggestion
	return fileUploadSessionID, t.teamDao.UpdateTeam(team)
```

File: core/service/team.go [hunk 6]
```diff
@@ -56,6 +65,87 @@ func (t Team) FindSprintsInTeam(ct context.Context, teamID uint64, filter *Sprin
...
+}
+
+func NewTeam(
+	cloudWebAPIExternalBaseURL string,
+	taskDao dao.Task,
+	sprintDao dao.Sprint,
+	teamDao dao.Team,
+	cloudClientRegistry *cloudAPI.ClientRegistry,
+	teamFileUploadSessionDao dao.TeamFileUploadSession) Team {
```
Comment (naturetime10) [inline core/service/team.go]:
```suggestion
	cloudClientRegistry *cloudAPI.ClientRegistry,
	taskDao dao.Task,
	sprintDao dao.Sprint,
	teamDao dao.Team,
	teamFileUploadSessionDao dao.TeamFileUploadSession,
) Team {
```

File: core/api/gql/team_mutation.go [hunk 7]
```diff
@@ -85,16 +86,42 @@ func (m Mutation) UpdateTeam(ct context.Context, args struct {
...
+		log.Println(err)
+		return "", err
+	}
+
+	uploadSessionID, err := m.deps.teamService.FinishTeamIconUploadSession(ct, teamID, fileUploadSessionID)
+	if err != nil {
+		return "", err
+	}
+
+	return graphql.ID(strconv.FormatUint(uploadSessionID, 10)), nil
```
Comment (naturetime10) [inline core/api/gql/team_mutation.go]:
```suggestion
	return toGraphQLID(uploadSessionID), nil
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: core/service/team.go (hunk 5 above)
Comment (duanxianbo) [reply core/service/team.go]: this should be an tuple of (id, error), right?

File: core/api/gql/team_mutation.go [hunk 8]
```diff
@@ -85,16 +85,42 @@ func (m Mutation) UpdateTeam(ct context.Context, args struct {
...
+	teamID, err := fromGraphQLID(args.TeamID)
+	if err != nil {
+		log.Println(err)
+		return "", err
+	}
+
+	uploadSessionID, err := m.deps.teamService.CreateTeamIconUploadSession(ct, teamID)
+	if err != nil {
+		return "", err
+	}
```
Comment (naturetime10) [inline core/api/gql/team_mutation.go]:
```suggestion
	if err != nil {
	    log.Println(err)
		return "", err
	}
```

File: core/api/gql/team_mutation.go [hunk 9]
```diff
@@ -85,16 +85,42 @@ func (m Mutation) UpdateTeam(ct context.Context, args struct {
...
+
+	teamID, err := fromGraphQLID(args.TeamID)
+	if err != nil {
+		log.Println(err)
+		return "", err
+	}
+
+	uploadSessionID, err := m.deps.teamService.FinishTeamIconUploadSession(ct, teamID, fileUploadSessionID)
+	if err != nil {
+		return "", err
```
Comment (naturetime10) [inline core/api/gql/team_mutation.go:122]:
```suggestion
        log.Println(err)
		return "", err
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

**Review by naturetime10 — APPROVED** [review]

**Review by naturetime10 — APPROVED** [review]

**Pull request merged**